### DISCUSSION TITLE
Multiple UARTs on the esp8266 / esp01?

### DISCUSSION THREAD

**Discussion by jouellnyc** [micropython/micropython#15280] [ESP8266] [answered]
Good day.

I have an ESP8266 / ESP-01S Board w/4MB flash. i.e https://www.amazon.com/dp/B08QF24GZZ

I thought this was interesting:

Note that I am aware of [dupterm](https://github.com/micropython/micropython/commit/afd0701bf7a9dcb50c5ab46b0ae88b303fec6ed3)

If I connect with Thonny via USB:

```
MicroPython v1.23.0 on 2024-06-02; ESP module with ESP8266
Type "help()" for more information.

>>> import machine
>>> uart = machine.UART(0, baudrate=9600, tx=1, rx=3)

# This hangs Thonny and makes perfect sense to me why

>>> uart = machine.UART(1, baudrate=9600, tx=1, rx=3)

# I would have thought this would yield a does not exist error?

>>> uart = machine.UART(2, baudrate=9600, tx=1, rx=3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: UART(2) does not exist

# This makes perfect sense to me
```

OK, well if pins 1 and 3 are used up for the REPL, and I am using the REPL, I guess only GPIO2 and 0 are left.

What If I try them?

```
>>> uart = machine.UART(1, baudrate=9600, tx=2, rx=0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid tx/rx

>>> uart = machine.UART(1, baudrate=9600, tx=0, rx=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid tx/rx

```

**What is my question? How many uarts can the esp01 use 'at once' in one program?**

It seems like it could be 2 based on creating a variable for UART1, but then not based on the pin combos.

p.s 

Just to try it and see what happens, I also put

```
uart = machine.UART(1, 115200)
os.dupterm(uart, 1)
```

in boot.py thinking maybe Thonny would then use that as the UART.

It never came up, which again would make sense to me but I thought I should try and note the results.

Thanks

**Comment by robert-hh** [comment, marked as answer]
The ESP8266 has in theory 2 UARTs available with ID 0 and 1. UART 1. The default pins are 1 and 3 for UART 0 and 2 and 8 for UART 1. Since Pin 8 is used internally, UART 1 is TX only. UART 0 can be swapped to use pins 15 (tx) and 13 (rx). 
See the documentation at https://docs.micropython.org/en/latest/esp8266/quickref.html#uart-serial-bus:
_Two UARTs are available. UART0 is on Pins 1 (TX) and 3 (RX). UART0 is bidirectional, and by default is used for the REPL. UART1 is on Pins 2 (TX) and 8 (RX) however Pin 8 is used to connect the flash chip, so UART1 is TX only._

**Comment by jouellnyc** [comment]
Firstly, thank you very much for the reply @robert-hh 

I wish my google foo got me to that page but it didn't. 

That's pretty awesome for know at least UART1 can at least  transmit.

Many thanks here!


For reference / FYI / For the curious, I tried to actually use the UART 1 on my ESP01s:

```
>>> uart = machine.UART(1, baudrate=9600, tx=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid tx/rx

>>> uart = machine.UART(1, baudrate=9600, tx=2,rx='NA')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid pin

>>> uart = machine.UART(1, baudrate=9600, tx=2,rx=8)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid pin

>>> uart = machine.UART(1, baudrate=9600, tx=8,rx=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid pin
```

Also, this is what happens when iterating through all the pins:

```

>>> for row in range(9):
  for col in range(9):
    #print(f"tx={col} rx={row}")
    try:
        uart = machine.UART(1, baudrate=9600, rx=row, tx=col)
    except ValueError:
        pass
    else:
        print(f"ok tx={col} rx={row}")
        
ok tx=1 rx=3

```

I suppose that perhaps other ESP01's boards may work better. 

But this was my experience with the above links from Amazon. 

I was trying to push this ESP01 as far as I could on one of my projects. I guess the mileage may vary.

**Reply by robert-hh** [reply]
Usually you do not have to specify rx and tx when using the default settings. So for UART 0, the tx=1 and rx=3, for UART 1, tx=2. To use UART1 just write e.g. `uart=UART(1, 9600)`
Then you will have output at Pin 2. The checks implemented in the software for proper tx/rx settings are misleading, sin ce these apply only to UART0 and the software does not take the UART number into account. So when using UART 1, tx is always at Pin 2. 
The tx/rx check could/should be fixed in the firmware, but given the number of pending PRs, this may not get a high priority.

**Reply by jouellnyc** [reply]
I had a feeling about the checks and I agree this is not exactly a high priority. 

Thanks for the syntax as well. I was so used to explicitly stating the pins (especially remapping /etc).

Finally, and again -- super awesome -- thanks @robert-hh
